modules/apps/frontend-js/frontend-js-a11y-web/src/main/resources/META-INF/resources/A11yChecker.ts

```diff
@@ -0,0 +1,422 @@
+		try {
+			return await this.workLoop();
+		}
+		finally {
+			this.currentTask = null;
+			this.isPerformingWork = false;
+		}
+	}
+
+	private requestHostCallback() {
+		this.scheduledHostHandle = window.requestIdleCallback(async () => {
+			const currentTime = performance.now();
+
+			// Defines a deadline only when the next phase of the execution
+			// frame arrives.
+
+			if (
+				this.nextFramePhase === 0 ||
+				currentTime >= this.nextFramePhase
+			) {
+				this.deadline = currentTime + this.yieldInterval;
+			}
+
+			const hasMoreWork = await this.flushWork();
+
+			if (hasMoreWork) {
+
+				// If there's more work, schedule the next callback at the
+				// end of the preceding one.
+
+				this.requestHostCallback();
+			}
+		});
+	}
+
+	private shouldYieldToHost(): boolean {
+		const currentTime = performance.now();
+
+		if (currentTime >= this.deadline) {
+
+			// We no longer have time to continue. We yield control of the main
+			// thread, so that the browser can perform high-priority tasks for
+			// a fixed period of time before starting to perform new tasks.
+
+			return this.nextFramePhase >= currentTime;
+		}
+
+		// There's still time left in the frame.
+
+		return false;
+	}
+
+	private async workLoop() {
+		this.currentTask = this.queue.peek();
+
+		while (this.currentTask !== null) {
+			if (this.shouldYieldToHost()) {
+
+				// We no longer have time to run currentTask now
+
+				break;
+			}
+
+			await this.currentTask.callback(this.currentTask.target);
+
+			this.queue.dequeue();
+
+			this.currentTask = this.queue.peek();
+
+			this.nextFramePhase = performance.now() + this.yieldInterval;
+		}
+
+		// Return whether there's additional work
+
+		return this.currentTask !== null;
+	}
+
+	scheduleCallback(callback: Function, target: T) {
+		const newTask = {
+			callback,
+			id: this.taskIdCounter++,
+			target,
+		};
+
+		this.queue.enqueue(newTask);
+
+		if (!this.isCallbackScheduled && !this.isPerformingWork) {
+			this.isCallbackScheduled = true;
+			this.requestHostCallback();
+		}
+	}
+}
+
+type Mutation = {
+	/**
+	 * Attributes are the attributes of a node. The array is considered a
+	 * conditional `or`, the same for the values of an attribute.
+	 */
+	attributes: Record<string, Array<string>>;
+
+	/**
+	 * NodeNames is an array of node names. This array is considered as a
+	 * conditional `or`.
+	 */
+	nodeNames: Array<string>;
+};
+
+export interface A11yCheckerOptions {
+
+	/**
+	 * AxeOptions is the set of settings that are passed to axe-core.
+	 */
+	axeOptions?: RunOptions;
+
+	/**
+	 * Callback for when the analysis is finished.
+	 */
+	callback: (violations: AxeResults) => void;
+
+	/**
+	 * Denylist is an optional list with CSS selectors that will be excluded
+	 * from the analysis.
+	 */
+	denylist?: Array<string>;
+
+	/**
+	 * Mutation is an optional list of criteria on which a new analysis will be
+	 * triggered.
+	 */
+	mutations?: Record<MutationRecordType, Mutation>;
```

Comment:
@dsanz I was thinking a little bit about this API and when I was doing some tests I find it a bit strange your behave, add criteria to validate a mutation I think it is more difficult for us to create a large list to validate various scenarios that are valid for a mutation I have in mind that it would be simpler to do the opposite, add criteria to ignore a mutation I see that it would be easier for me to ignore some mutations than having to know how the mutation behaviors of part of my application and to add within a criterion. Does it make sense?

Reply:
It does yeah. The idea behind this is basically to be able to ignore 2 sets of mutations 
* The ones our a11y checker tool produces: the popovers, side panel, detail pop-ups... We don't want to analyze how accessible our panels are. Well, we may want to do it, but only when we develop the checker itself (kinda "dev mode within dev mode"), not when the a11y checker is helping liferay devs to build accessible features.
* When integrated in DXP, we'll like to ignore the mutations produced by typing into text fields/areas such as the ones provided by any of the editors we support

So, it's likely we want to implement a "blacklist" mutation criteria and assume everything else will be enqueued for analysis

Reply:
Hmmm, So, I understand that the mutation criteria are used to trigger a new analysis when a read the ticket, now I understand that mutation criteria are used to ignore one mutation. 

I'm a little confused now, so the mutations API will be used internally and not exposed (the user can define), right? and a mutation criterion is used to ignore mutation and not the other way around, right?

Reply:
> Hmmm, So, I understand that the mutation criteria are used to trigger a new analysis when a read the ticket, now I understand that mutation criteria are used to ignore one mutation.

I think we're talking about the same thing, just from 2 different perspectives. The requirement is to make the tool able to selectively act upon mutations, meaning that, from the full set of mutations that can take place on the page, only a subset of them would trigger a new analysis. This is expected to save some cpu and also to leave room for application-level adaptations.

This requirement can be implemented in many ways, I can think of 2 primary approaches:
- **Blacklist**: any mutation matching the blacklist criteria will not be scheduled for analysis
- **Whitelist**. only mutations matching the whitelist criteria will be scheduled for analysis
 
In light of previous conversations, it looks like a blacklist approach would be more beneficial as the amount of work to define it seems smaller and the maintenance, easier.

There is an additional advantage of using a blacklist, which I did not consider before. Let's suppose there is a mutation which does not match with the blacklist. When scheduling the mutation for analysis, we can feed the Task with the mutation criteria as part of the axe include-exclude object, so we'll prevent axe to analyze certain parts of the mutation which we're not interested in. This is something we don't need to consider in this iteration, and may even not make any sense if turning mutation criteria into include-exclude objects is not a cheap operation. Still, this can be something to consider.

> I'm a little confused now, so the mutations API will be used internally and not exposed (the user can define), right? and a mutation criterion is used to ignore mutation and not the other way around, right?

It depends on what do we consider user. When we address the integration with DXP, we'll find out which criteria make sense, and define them in the a11y configuration setting for DXP. In this sense, we are the users and we'll need the tool to expose a configuration API to define the mutations. This can be as easy as a js function, no need to have anything ellaborated.

Analogously, portal admins may like to tweak these DXP defaults in case they want to analyze the a11y for custom developments, so there are chances that the JS function would be something we'd like to be provided. I don't think that end-users should have access to this configuration API.

Reply:
I think I understand, I will leave only a vision of how it is being done in this PR so we can see what we have so far.

We have two main APIs that can help us save CPU resources:

- **Denylist** is an optional list with CSS selectors that will be excluded from the analysis. We just move on to `axe-core.`
- **Mutations** is an optional list of criteria on which a new analysis will be triggered.

> I'm using that term instead of blacklist which is being somewhat abolished from the codebase and a way to avoid racism...

We can think of the `mutations` API as a middleware, which acts at the stage when a DOM mutation occurs and we use it as an `allowlist` to check if we should schedule the analysis. So the idea that I think we agree that we should make the `mutations` API as a `denylist` because it will be easier to configure and maintain. Does It make sense?

Reply:
Thanks for the explanations Matu. To begin, I support the avoidance of racist terms, so I'll change the terminology right away, and also changed the LPS texts.

Yes, the idea would be to define the mutations as a denylist too. In the end, `axe-core` allows both allow and deny lists for analysis via include-exclude object. As you explain, we are adding an extra layer which controls the invocations of `axe-core` via mutation observation, and that layer does not have a counterpart in the `axe-core` api afaik. So this one would only exist as a denylist as it better suites our use cases.

The expected use case is to avoid triggering mutations which results from specific user interactions such as typing, clicking on a checkbox/radio button, selecting an item from a dropdown, etc. These are assumptions made on the basis that what we're not interested on the a11y violations produced by such mutations in the DOM. Nonetheless, if any of such actions produces an additional change in the DOM such as rendering a pop-up, that one will be a separate mutation which will not be in the denylist and so will be analyzed.

Does that make sense?

Reply:
Hmm okay. So, I'm going to change the `mutations` to the behavior of a denylist.

Hmm, I think I understood your reasoning about this, well we didn’t reach that level of knowing what caused the mutation, following these examples, click, typing (typing in an input does not cause a mutation, it is disabled by default), selecting something, but only the final result of these actions that can occur in DOM changes, so what we know is if the element has been removed, added according to the target and if any attributes have changed. Our API will look like this:

```js
const mutations = {
  // MutationRecordType
  attributes: {
    attributes: {
      // This is an `or` operator
      class: ['show', 'hide']
    },
    // This is an `or` operator
    nodeNames: ['li']
  },
};
```

I'm still not sure if I'm going to continue with this API, it doesn't seem very flexible yet, seeing ways on how to approach it better.

When a mutation becomes valid we schedule an analysis run, then we use the `include` and `exclude` object to pass to axe-core, so ignore our `denylist` from the analysis and we use the `include` to be the target element for the analyze.

Reply:
> Hmm, I think I understood your reasoning about this, well we didn’t reach that level of knowing what caused the mutation, following these examples, click, typing (typing in an input does not cause a mutation, it is disabled by default), selecting something, but only the final result of these actions that can occur in DOM changes, so what we know is if the element has been removed, added according to the target and if any attributes have changed. 

I've making some tests, and according to the results I got, I think we can simplify this a lot. The thing is that, contrary to my initial belief, the interactions one makes with things like text inputs, selects, radios and check buttons are not producing DOM mutations, meaning that all changes produced on those are not part of the dom. This [old chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=128991) clarifies what I mean. My initial belief was that we should configure the mutation observer in order to ignore these kind of mutations caused by direct user interactions on the UI controls.

Considering the tests I made, only the `contenteditable="true"` on an element seems to produce observable mutation events when typing. This reduces the amount of things we must take care of as most of them are already _unobservable_ by the mutation observer api.

Said that, perhaps the only thing we should support is the possiblity of supplying a [MutationObserverInit](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserverInit) object to the checker, which would control which mutations are to be observed in order to trigger analysis. I believe this should be enough, but please let me know if you think otherwise.

> When a mutation becomes valid we schedule an analysis run, then we use the `include` and `exclude` object to pass to axe-core, so ignore our `denylist` from the analysis and we use the `include` to be the target element for the analyze.

In this case, I think that supplying the `denylist` could also be useful. I'm thinking on a case such as the page editor, where double-clicking creates an editor on the fly. Such a mutation may trigger an analysis, but we'd like to ignore what's inside the editor. So the set of nodes we want to analyze is not exactly the same as the added mutations and its desdendants. Does this make sense?

Reply:
> Said that, perhaps the only thing we should support is the possiblity of supplying a MutationObserverInit object to the checker, which would control which mutations are to be observed in order to trigger analysis. I believe this should be enough, but please let me know if you think otherwise.

I think it might make sense, the current API implemented would offer a better granularity for the mutation filter but I'm not sure if it's a good thing that we want to endure sometimes I have a feeling that this can be bad because we are giving it the possibility to ignore things from the analysis that we can hide accessibility errors in the way it is implemented, so I'm fine with just supporting `attributeFilter` to ignore mutations like `contenteditable`.

We currently define `MutationObserverInit` with:

- `attributes` true
- `childList` true
- `subtree` true

I think if we need something more flexible we can implement it in the future and maybe we can design a better API, since we can have a better view of what we want to allow.

> In this case, I think that supplying the denylist could also be useful. I'm thinking on a case such as the page editor, where double-clicking on a paragraph creates an editor on the fly. Such a mutation may trigger an analysis, but we'd like to ignore what's inside the editor. So the set of nodes we want to analyze is not exactly the same set of added mutations (and its desdendants) as reported by the observer api. Does this make sense?

Yeah, makes sense. Currently, I have implemented it this way.

---

Well, I think we're aligned! So I'm going to change the API from `mutations` or just to `mutation` to support `attributeFilter`, does it make sense, just to support this property? I think the other properties we can leave as is because it allows the checker to work better, guaranteeing to analyze new parts of the UI that have been added, for example in a panel change.

Reply:
> Actually, I'm not sure if the `attributeFilter` is ideal here to filter mutations because it has the behavior of an `allowlist`, I'm just thinking about simplifying the `mutations` API to filter the `attributes` and the ` nodeNames` or just keep it as is and make mutations a denylist, what do you think?

I wish I had a crystal ball to tell you what's the best option. I'd say there is a tradeoff between how complex our mutation observers are and how many "useless analysis" we trigger, also, how "big" those analysis are in terms of #of nodes and #of rules in the `axe.run` call. 

Initially, I'd lean towards a deny list as it's simpler to configure and less prone to leave things out. Now, we're interested in seeing how well the tool catches as many violations as possible. In light of this, we may start with a coarse-grain solution (observe all attrs, with all rules)

Nonetheless, I think there are open questions, even beyond the nature of the mutation list api (allow-ish vs deny-ish):
- How different combinations of `attributeFilter` "native" allowlist may impact performance, in comparison with an equivalent denylist implemented on top of the standard mutation api, which would discard certain mutations already reported (time for the native api to report vs time to filter the reported results)
- Which attributes are used by the [axe-core checks](https://github.com/dequelabs/axe-core/blob/master/lib/checks) to figure it out which rules requre the observation of certain attributes. For instance, the [color contrast check](https://github.com/dequelabs/axe-core/blob/master/lib/checks/color/color-contrast-evaluate.js) may be a good fit whenever the mutated attributes include things like `style` or `class`. This may be an argument in favor of allowlist.
- Fine tune which observations will trigger which checks (rather than checking for a fixed set of rules). For instance, a change in attributes related to style will not trigger rules related to heading order.
- Fine tune which nodes should be added to the analysis (analyze always the root target, or only the added nodes and the parent of the removed nodes)
- Fine tune the scope of the analysis to some area of interest, like for instance discarding the common elements like product menu, etc

All that leads me to believe that we shall consider some degree of internal configuration for the purpose of allowing this exploration, which would not necessarily be exposed to the checker users/integrators. Again, that's something I would not care about much in this iteration unless the performance becomes really bad.

Reply:
They are great observations and I totally agree. I think these are things that we will probably discover when we actually integrate this into DXP and see people interacting, currently, this is very performatic but it can be slow at times, for example, there is a cost for the analysis when the tree is very large, so granular analysis is certainly very beneficial here, despite having a "fiber" based implementation to avoid thread blocking and yield for everyone, there are some problems:

- Having too many tasks scheduled is not the problem, the problem is how long the scheduler expects the task to be completed to run the next task at the best time
- The analysis is around 40ms ~ 200ms, 40ms usually in single elements sometimes, so I think it makes a lot of sense here when you say it may not make sense to add all the checks, just because of a change in style...
- When there are many scheduled tasks where analysis can take a long time, the developer will only notice a certain slowness of the overlays appearing after the page has been rendered but it does not affect the performance of the page or prevent the user from continuing to interact with the page

The point of the scheduler is that I think we are doing well with this implementation but the best here would be to have analyzes being executed in less time, so we can have the best of all worlds. Mutation validations do not seem to be a big problem, as they are not expensive in terms of time.

I think to summarize a little and to be able to align what we have now is:

- `denylist` for the axe-core that ignores elements of the analysis (It also helps to shorten the analysis time)
- `mutations` to ignore mutations in the DOM in order to avoid unnecessary analysis (does not configure by default, we can configure when integrating with DXP)

I like the idea of defining the checks according to the mutation change, this can greatly reduce the analysis time, maybe something that we can explore in future interactions. So considering your observations a little bit, I think it would be good to try to explore that at some point.

I think we can reduce the analysis time more considering:

- `checks` adjustable according to the type/node of mutation
- We continue to analyze the element that triggered the analysis instead of analyzing the target again